lib/zlib.c

```diff
@@ -0,0 +1,493 @@
+			goto out;
+		}
+
+		lvl = (int)ucv_int64_get(level);
+	}
+
+	ret = deflateInit2(&strm, lvl,
+			   Z_DEFLATED,		// only allowed method
+			   gz ? 15+16 : 15,	// 15 Zlib default, +16 for gzip
+			   8,			// default value
+			   Z_DEFAULT_STRATEGY);	// default value
+	if (ret != Z_OK) {
+		uc_vm_raise_exception(vm, EXCEPTION_RUNTIME, "Zlib error: %s", ziniterr(ret));
+		goto out;
+	}
+
+	switch (ucv_type(src)) {
+	case UC_STRING:
+		buf = uc_zlib_def_string(vm, src, &strm);
+		break;
+
+	case UC_RESOURCE:
+	case UC_OBJECT:
+	case UC_ARRAY:
+		buf = uc_zlib_def_object(vm, src, &strm);
+		break;
+
+	default:
+		uc_vm_raise_exception(vm, EXCEPTION_TYPE,
+				      "Passed value is neither a string nor an object");
+		goto out;
+	}
+
+	if (!buf) {
+		if (vm->exception.type == EXCEPTION_NONE)	// do not clobber previous exception
+			uc_vm_raise_exception(vm, EXCEPTION_RUNTIME, "Zlib error: %s", strm.msg);
+		goto out;
+	}
+
+	rv = ucv_stringbuf_finish(buf);
+
+out:
+	(void)deflateEnd(&strm);
+	return rv;
+}
+
+static uc_stringbuf_t *
+uc_zlib_inf_object(uc_vm_t *vm, uc_value_t *obj, z_stream *strm)
+{
+	int ret;
+	bool eof = false;
+	uc_value_t *rfn, *rbuf;
+	uc_stringbuf_t *buf = NULL;
+
+	rfn = ucv_property_get(obj, "read");
+
+	if (!ucv_is_callable(rfn)) {
+		uc_vm_raise_exception(vm, EXCEPTION_TYPE,
+				      "Input object does not implement read() method");
+		return NULL;
+	}
+
+	buf = ucv_stringbuf_new();
+
+	do {
+		rbuf = NULL;
+		uc_vm_stack_push(vm, ucv_get(obj));
+		uc_vm_stack_push(vm, ucv_get(rfn));
+		uc_vm_stack_push(vm, ucv_int64_new(CHUNK));
+
+		if (uc_vm_call(vm, true, 1) != EXCEPTION_NONE)
+			goto fail;
+
+		rbuf = uc_vm_stack_pop(vm);	// read output chunk
+
+		/* we only accept strings */
+		if (rbuf != NULL && ucv_type(rbuf) != UC_STRING) {
+			uc_vm_raise_exception(vm, EXCEPTION_TYPE,
+					      "Input object read() method returned non-string value");
+			goto fail;
+		}
+
+		/* check EOF */
+		eof = (rbuf == NULL || ucv_string_length(rbuf) == 0);
+		if (eof)
+			break;
+
+		strm->next_in = (unsigned char *)ucv_string_get(rbuf);
+		strm->avail_in = ucv_string_length(rbuf);
+
+		/* run deflate() on input until output buffer not full */
+		do {
+			// enlarge buf by CHUNK amount as needed
+			printbuf_memset(buf, printbuf_length(buf) + CHUNK - 1, 0, 1);
+			buf->bpos -= CHUNK;
+
+			strm->avail_out = CHUNK;
+			strm->next_out = (unsigned char *)(buf->buf + buf->bpos);;
+
+			ret = inflate(strm, Z_NO_FLUSH);
+			assert(ret != Z_STREAM_ERROR);		// state never clobbered (would be mem corruption)
+			switch (ret) {
+			case Z_NEED_DICT:
+			case Z_DATA_ERROR:
+			case Z_MEM_ERROR:
+				goto fail;
+			}
+
+			// update bpos past data written by deflate()
+			buf->bpos += CHUNK - strm->avail_out;
+		} while (strm->avail_out == 0);
+
+		ucv_put(rbuf);	// release rbuf
+	} while (ret != Z_STREAM_END);	// done when inflate() says it's done
+
+	if (ret != Z_STREAM_END) {	// data error
```

Comment:
The static clang analyzer found a mistake here;
when the read callback returns EOF on the first loop iteration, we break out before setting the uninitialized `ret` to anything. We probably should initialize `int ret` to `Z_STREAM_END` in line 282.

Reply:
I think I would initialize as `Z_STREAM_ERROR` instead since passing an empty input is a stream error after all and it avoids returning a buffer with random garbage. OK?

Reply:
I guess that would work too. Reason I was suggesting `Z_STREAM_END` is that it would skip the subsequent `printbuf_free(buf)` call (which is null safe, but `buf` is guaranteed to be `NULL` in this situation). Since we do not return `ret` in any way to the caller or user, it does not really matter.

But either way will work.

Reply:
Well if we EOF in the first loop, buf *has been* allocated (line 295) so it makes sense to free it and return `NULL` instead, IMHO. Will push change